libcaf_core/caf/action.hpp

```diff
@@ -132,34 +136,87 @@ inline bool operator!=(const action& lhs, const action& rhs) noexcept {
 namespace caf::detail {
 
 template <class F, bool IsSingleShot>
-struct default_action_impl : detail::atomic_ref_counted, action::impl {
-  std::atomic<action::state> state_;
-  F f_;
-
+class default_action_impl : public detail::atomic_ref_counted,
+                            public action::impl {
+public:
   default_action_impl(F fn)
     : state_(action::state::scheduled), f_(std::move(fn)) {
     // nop
   }
 
+  ~default_action_impl() {
+    // The action going out of scope can't be running or deferred dispose.
+    CAF_ASSERT(state_.load() != action::state::running);
+    CAF_ASSERT(state_.load() != action::state::deferred_dispose);
+    if (state_.load() == action::state::scheduled)
+      f_.~F();
+  }
+
   void dispose() override {
-    state_ = action::state::disposed;
+    // Try changing the state to disposed
+    while (true) {
+      if (auto expected = state_.load(); expected == action::state::scheduled) {
+        if (state_.compare_exchange_weak(expected, action::state::disposed)) {
+          f_.~F();
+          return;
+        }
+      } else if (expected == action::state::running) {
+        if (state_.compare_exchange_weak(expected,
+                                         action::state::deferred_dispose))
+          return;
+      } else { // action::state::{deferred_dispose, disposed}
+        return;
+      }
+    }
+  }
+
+  void handle_deferred_dispose() {
   }
 
   bool disposed() const noexcept override {
-    return state_.load() == action::state::disposed;
+    auto current_state = state_.load();
+    return current_state == action::state::disposed
+           || current_state == action::state::deferred_dispose;
   }
 
   action::state current_state() const noexcept override {
     return state_.load();
   }
 
+  void run_single_shot() {
+    // We can only run a scheduled action
+    auto expected = action::state::scheduled;
+    if (!state_.compare_exchange_strong(expected, action::state::disposed))
+      return;
+    f_();
+    f_.~F();
+    return;
+  }
+
+  void run_multi_shot() {
+    // We can only run a scheduled action
+    auto expected = action::state::scheduled;
+    if (!state_.compare_exchange_strong(expected, action::state::running))
+      return;
+    f_();
+    // Once run, we can stay in the running state or switch to deferred dispose.
+    expected = state_.load();
+    if (expected == action::state::running
+        && state_.compare_exchange_strong(expected, action::state::scheduled))
+      return;
+    if (expected == action::state::deferred_dispose
+        && state_.compare_exchange_strong(expected, action::state::disposed)) {
+      f_.~F();
+      return;
+    }
+    CAF_RAISE_ERROR("Invalid state found after running an action.");
```

Comment:
I see what you did there. I agree with the changes, but I think without the line:
`auto expected = state_.load();`
expected would be left as `action::state::scheduled`, since compare exchange does nothing to expected if it succeeds.